05-JWT-Basics/starter/middleware/auth.js

```diff
@@ -0,0 +1,20 @@
+const jwt = require("jsonwebtoken");
+const { UnauthenticatedError } = require("../errors");
+
+const authenticationMiddleware = async (req, res, next) => {
+  const authHeader = req.headers.authorization;
+  if (!authHeader || !authHeader.startsWith("Bearer ")) {
+    console.log("NO TOKEN ! ");
+    throw new UnauthenticatedError("No Token Provided");
+  }
+  const token = authHeader.split(" ")[1];
+  try {
+    const decoded = jwt.verify(token, process.env.JWT_SECRET);
+    const { id, username } = decoded;
+    req.user = { id, username };
+  } catch (error) {
+    throw new UnauthenticatedError("Not authorized");
+  }
+  next();
```

Comment:
Nitpick:
This can token validation logic can be simplified using [optional chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining). We don't really care if the header and token is there, instead we only really care if the header token is valid!
```js
// optional chaining allows you to access a property that may or may not be there
// if it's not there, it'll fall back to undefined
const token = req.headers.authorization?.split(" ")?.[1];

try {
  const user = jwt.verify(token, process.env.JWT_SECRET);
  req.user = user;
  next();
} catch (error) {
  throw new UnauthenticatedError("Not authorized");
}
```